orca_detector/generator.py

```diff
@@ -1,94 +1,180 @@
+"""
+
 import keras
 import librosa
 import numpy as np
+import pprint
 import resampy
 import soundfile as sf
 
-import params
-from vggish.mel_features import frame, log_mel_spectrogram
-
+# project-specific imports
+import orca_params
+import vggish_params as params
+from mel_features import frame, log_mel_spectrogram
 
 def _waveform_to_examples(data, sample_rate):
-  # Convert to mono.
-  if len(data.shape) > 1:
-    data = np.mean(data, axis=1)
-  # Resample to the rate assumed by VGGish.
-  if sample_rate != params.sample_rate:
-    data = resampy.resample(data, sample_rate, params.sample_rate)
-
-  # Compute log mel spectrogram features.
-  log_mel = log_mel_spectrogram(
-      data,
-      audio_sample_rate=params.sample_rate,
-      log_offset=params.log_offset,
-      window_length_secs=params.stft_window_length_seconds,
-      hop_length_secs=params.stft_hop_length_seconds,
-      num_mel_bins=params.num_mel_bins,
-      lower_edge_hertz=params.mel_min_hz,
-      upper_edge_hertz=params.mel_max_hz )
-
-  # Frame features into examples.
-  features_sample_rate = 1.0 / params.stft_hop_length_seconds
-  example_window_length = int(round(
-      params.quantum_size* features_sample_rate))
-  example_hop_length = int(round(
-      params.quantum_hop * features_sample_rate))
-  log_mel_examples = frame(
-      log_mel,
-      window_length=example_window_length,
-      hop_length=example_hop_length)
-  return log_mel_examples
+    """
+    Converts audio from a single wav file into an array of examples for VGGish.
+
+    Args:
+        data: np.array of either one dimension (mono) or two dimensions
+          (multi-channel, with the outer dimension representing channels).
+          Each sample is generally expected to lie in the range [-1.0, +1.0],
+          although this is not required. Shape is (num_frame, )
+        sample_rate: Sample rate of data.
+
+    Returns:
+        3-D np.array of shape [num_examples, num_frames, num_bands] which represents
+        a sequence of examples, each of which contains a patch of log mel
+        spectrogram, covering num_frames frames of audio and num_bands mel frequency
+        bands, where the frame length is params.STFT_HOP_LENGTH_SECONDS.
+
+    IMPORTANT: if data.shape < (80000, ) then log_mel_examples.shape=(0, 496, 64).
+        The zero is problematic downstream, so code will have to check for that.
+    """
+    
+    # Convert to mono if necessary.
+    if len(data.shape) > 1:
+        print('DEBUG: audio channels before={}'.format(data.shape))
+        data = np.mean(data, axis=1)
+        print('DEBUG: audio channels after={}'.format(data.shape))
+        
+    # Resample to the rate assumed by VGGish.
+    if sample_rate != params.SAMPLE_RATE:
+        data = resampy.resample(data, sample_rate, params.SAMPLE_RATE)
+
+    # Compute log mel spectrogram features.
+    log_mel = log_mel_spectrogram(data,
+                                  audio_sample_rate=params.SAMPLE_RATE,
```

Comment:
I wanted to decouple the VGGish code dependency from our code in case we want to try something else, this is why I created a facade of the params. Also, by doing that we would be able to dynamically override our params in ENV files or docker settings. This is also why I copied this code from the VGGish example

Reply:
The generation of the log mel spectrogram is decoupled from the VGGish model.  Pretty much every example I saw of people doing audio classification relied on first generating the log mel spectrogram to feed into their model as features.  However, you have a point that they don't really belong in the vggish_params.py file.  I'll split the appropriate ones into a mel_params.py file so that we can have 3 total param files:
- mel_params.py will have params related to generation of the spectrograms
- vggish_params.py will have params specific to that model
- orca_params.py will have other misc params for our project (e.g. paths, batch size)